### ISSUE TITLE
Client-side data portal needs to Dispose proxy instances

### ISSUE THREAD

**Issue by TheCakeMonster** [MarimerLLC/csla#2946]
The viral effect of async/await continues apace ...

I have implemented IAsyncDisposable on a class that I am injecting into a data access method of a CSLA class using the [Inject] attribute. An exception is raised when a data portal operation completes. I think this is because the data portal is cleaning up the scope that it creates, but doesn't support async disposable classes - it only offers synchronous cleanup.

```
System.InvalidOperationException: 'DotNotStandard.DataAccess.ConnectionManager' type only implements IAsyncDisposable. Use DisposeAsync to dispose the container.
 at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
 at Csla.Channels.Local.LocalProxy.Dispose(Boolean disposing)
 at Csla.Channels.Local.LocalProxy.Dispose()
 at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.DisposeAsync()
--- End of stack trace from previous location ---
 at Microsoft.AspNetCore.Http.Features.RequestServicesFeature.<DisposeAsync>g__Awaited|9_0(RequestServicesFeature servicesFeature, ValueTask vt)
 at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<FireOnCompleted>g__ProcessEvents|227_0(HttpProtocol protocol, Stack`1 events)
```

I am using LocalProxy in this particular case which is likely to be the one most affected. However, this might also apply to any other data portal implementations if they too create their own scopes.

**Proposed Solution**
Implement IAsyncDisposable on LocalProxy, and any other proxy that needs to perform the same clean up as a result of injecting services and/or creating a scope.

**Alternatives Considered**
My class implements only IAsyncDisposable and not IDisposable. I have thought about implementing IDisposable as well, but in the case of the class I am creating it is difficult to implement the cleanup operation synchronously without risking deadlocking - the underlying classes that might be used are likely to only offer async methods to perform the cleanup required.

**Additional context**
I'm using CSLA 6 and a .NET 6 Blazor Server project with the default Blazor Server config for CSLA, and thus LocalProxy.

**Comment by TheCakeMonster** [comment]
> This is new stuff to me.

Yeah, me too TBH.

> So you have DI create your service, and when the scope is disposed, the scope isn't automatically disposing your service?

When the scope is disposed it's _trying_ to dispose of my service automatically. However, it is failing, because it needs disposing asynchronously, and we don't have quite the right support for that.

> We are thinking that `LocalProxy`, when it disposes the scope, needs to asynchronously dispose the scope?
> 
> I don't see how making `LocalProxy` disposable (sync or async) would help, unless the client-side data portal did the same. And if that happened, then the calling code (often your UI code) would need to do the top-level disposing?

You raise an interesting point. I didn't think it would be necessary to spread it all the way to the client - but I have to confess it's fairly new to me, and I'm not entirely sure either.

I _thought_ what we'd need to do is update the four data access methods in `DataPortal<T>` (which are already async) to support async disposal of the proxy that is created by the call to `GetDataPortalProxy`.

As it happens, I don't see them cleaning up the proxy if they created it anyway - there is no call to Dispose on the proxy - is that a bug/imperfection? I think it might be.

It appears that the proxy isn't always created in the way I expected. `GetDataPortalProxy` doesn't always create an instance itself. However, when it doesn't, I think it is created by DI and passed to the constructor of `DataPortal<T>`, so as long as the proxy implementation supports `IAsyncDisposable` then I think the container would probably clean it up for us in that case.

I think that maybe the following changes would be required:
1. Add `IAsyncDisposable` support to the proxy implementations (this is just boilerplate code.)
2. Add `IAsyncDisposable` to `DataPortal<T>` (the exact same boilerplate code.)
3. Create an async method in `DataPortal<T>` called `DisposeProxyAsync` containing fairly similar code. This simply uses an if statement to decide whether to call Dispose or await DisposeAsync.
4. Call `DisposeProxyAsync`, passing the proxy object, from each of the `DoCreateAsync`, `DoFetchAsync`, `DoUpdateAsync` and `DoDeleteAsync` methods.

Steps 1 and 2 alone might fix it in most cases - perhaps the scenario I have failing at the moment. If the proxy is injected by the container then the container deals with disposal automatically, and the container already supports `IAsyncDisposable`. However, I think there is subtlety about whether the proxy gets created by the container automatically or whether the `GetDataPortalProxy` method does it that we'd need to take care of - hence steps 3 and 4.

https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync

**Comment by rockfordlhotka** [comment]
There's clearly a bug, in that `proxy` is never disposed at all. For example, there should be a `finally` block [here](https://github.com/MarimerLLC/csla/blob/d995e0b937d666b4e863460a687857581a026e07/Source/Csla/DataPortalT.cs#L437) to dispose `proxy`.

I don't know if we need to do anything about the async or not - but surely `Dispose` should be called for every proxy instance created.

**Comment by TheCakeMonster** [comment]
The async stuff would be a nice addition, in my opinion. I can see other people running into this.

It's not hard to do I don't think. I've added a check for this in one of my libraries today.

The only slight peculiarity is that it looks like you might need to reference the Microsoft.Bcl.AsyncInterfaces package from a .NET Standard 2.0 library. I was getting a runtime error from a unit test project running on .NET 6 but using a .NET Standard 2.0 library until I added this reference to that library - despite it compiling fine without the reference.
